File: py-project/Solution43.py
# -*- coding:utf-8 -*-
class Solution:
    def NumberOf1Between1AndN_Solution(self, n):
        # write code here
        if n<=0:
            return 0
        num = 0
        i = 1
        while i<=n:
            div = i*10
            k = n%div
            if k> i*2-1:
                c = i
            elif k<i:
                c = 0
            else:
                c = k-i+1
            num += (n//div)*i+c
            i = i*10
        return  num

if __name__ == '__main__':
    # l = range(1,12)
    # print(l)
    l = 12
    solution1 = Solution()
    a = solution1.NumberOf1Between1AndN_Solution(l)
    print(a)

File: c2-project/Solution43.cpp
//
// Created by 陈高杰 on 2020/2/11.
//
#include <iostream>
#include <stdlib.h>
#include <string>
#include <vector>
#include <stack>
#include <queue>
using namespace std;

class Solution {
public:
    int NumberOf1Between1AndN_Solution(int n)
    {
        if(n<=0){
            return 0;
        }
        int num = 0;
        for (int i = 1; i <= n ; i *=10) {
            int div = i*10;
            int k = n%div;
            int c = 0;
            if(k>2*i-1){
                c = i;
            }
            else if (k<i){
                c = 0;
            }
            else{
                c = k-i+1;
            }
            num += (n/div)*i+c;
        }
        return num;
    }
};
//
//int main(){
//    int l = 12;
//    Solution solution;
//    int x = solution.NumberOf1Between1AndN_Solution(l);
//    cout << x<< endl;
//    system("PAUSE");
//    return 0;
//
//}
